File: dp/509.java
class Solution 
{
    Map<Integer,Integer> memo = new HashMap<Integer,Integer>();
    public int fib(int n) 
    {
        
        if(n == 0 || n ==1) return n;
        if(memo.containsKey(n))
        {
            return memo.get(n);
        }

        int result = fib(n-1) + fib(n-2);

        memo.put(n,result);

        return result;
        
    }
}

File: 509.cpp
class Solution 
{
  public:
    unordered_map<int,int> dp;
    int fib(int n) 
    {
      if(n==0) return 0;
      if(n==1) return 1;
      if(dp.count(n)) return dp[n];

      else{
        return dp[n]=fib(n-1)+fib(n-2);

      }
    }
};
